Tests/SonarScanner.MSBuild.PreProcessor.Test/WebClientDownloaderTest.cs

```diff
@@ -146,26 +150,156 @@ public async Task Download_Fail_NotForbidden_WithLog()
+        }
+
+        [TestMethod]
+        public async Task TryDownloadIfExists_HttpCodeNotFound_SucceedWithLog()
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.NotFound);
+
+            var result = await sut.TryDownloadIfExists(RelativeUrl);
+
+            result.Item1.Should().BeFalse();
+            result.Item2.Should().BeNull();
+            logger.AssertDebugLogged($"Downloading from {BaseUrl}{RelativeUrl}...");
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public async Task TryDownloadIfExists_HttpCodeForbidden_ShouldThrowWithLog()
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.Forbidden);
+
+            Func<Task> act = async () => await sut.TryDownloadIfExists(RelativeUrl);
+
+            await act.Should().ThrowExactlyAsync<HttpRequestException>();
+            logger.AssertDebugLogged($"Downloading from {BaseUrl}{RelativeUrl}...");
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public async Task TryDownloadIfExists_HttpCodeForbiddenWithWarnLog_ShouldThrowWithLog()
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.Forbidden);
+
+            Func<Task> act = async () => await sut.TryDownloadIfExists(RelativeUrl, true);
+
+            await act.Should().ThrowExactlyAsync<HttpRequestException>();
+            logger.AssertDebugLogged($"Downloading from {BaseUrl}{RelativeUrl}...");
+            logger.AssertWarningLogged("To analyze private projects make sure the scanner user has 'Browse' permission.");
+        }
+
+        public static IEnumerable<object[]> UnsuccessfulHttpCodeData =>
+            Enum.GetValues(typeof(HttpStatusCode))
+                .Cast<HttpStatusCode>()
+                .Where(x => ((int)x >= 300 || (int)x < 200) && x != HttpStatusCode.Forbidden && x != HttpStatusCode.NotFound) // Those have specific tested behavior
+                .Select(x => new object[] { x });
+
+        [TestMethod]
+        [DynamicData(nameof(UnsuccessfulHttpCodeData))]
+        public async Task TryDownloadIfExists_UnsuccessfulHttpCode_ShouldThrowWithLog(HttpStatusCode code)
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, code);
+
+            Func<Task> act = async () => await sut.TryDownloadIfExists(RelativeUrl);
+
+            await act.Should().ThrowExactlyAsync<HttpRequestException>();
+            logger.AssertDebugLogged($"Downloading from {BaseUrl}{RelativeUrl}...");
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public async Task TryDownloadFileIfExists_HttpCodeOk_SucceedWithLog()
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.OK);
+            var fileName = Path.GetRandomFileName();
```

Comment:
We create the file in test but we never delete it. This can pollute  a bit our dev machines. Not sure how much of a problem it is but we could delete it at the end of the test.

Reply:
Indeed, I didn't realize that!
While fixing it, I saw some `TestUtilities` create folders that are not removed by the end of a test session. It may be worth looking into it to see if we also want to fix this.